src/log.rs

```diff
@@ -55,12 +67,50 @@ impl<T: Transport> Stream for LogStream<T> {
             #[project]
             let next_state = match state.as_mut().project() {
                 LogStreamState::CreatingFilter(create_filter, poll_interval) => {
-                    let log_filter = match ready!(create_filter.poll(cx)) {
-                        Ok(log_filter) => log_filter,
+                    let filter = match ready!(create_filter.poll(cx)) {
+                        Ok(filter) => filter,
                         Err(err) => return Poll::Ready(Some(Err(err.into()))),
                     };
-                    let stream = log_filter.stream(*poll_interval).compat();
-                    LogStreamState::Streaming(stream)
+                    // TODO: this request can be very slow and the response very big. We might want
+                    // to be smarter about it. However, there is no way to page this request. Maybe
+                    // Set up intermediate queries to handle a limited number of blocks at a time.
+                    let past_logs_future = CompatCallFuture::<T, Vec<Log>>::new(filter.logs());
+                    LogStreamState::GettingPastLogs {
+                        poll_interval: *poll_interval,
+                        filter,
+                        past_logs_future,
+                    }
+                }
+                LogStreamState::GettingPastLogs {
+                    poll_interval,
+                    filter,
+                    past_logs_future,
+                } => {
+                    let mut past_logs = match ready!(past_logs_future.poll(cx)) {
+                        Ok(past_logs) => past_logs,
+                        Err(err) => return Poll::Ready(Some(Err(err.into()))),
+                    };
+                    // Reverse so that we can pop the oldest event from the end.
+                    past_logs.reverse();
+                    LogStreamState::StreamingPastLogs {
+                        poll_interval: *poll_interval,
+                        filter: filter.clone(),
+                        past_logs,
+                    }
+                }
+                LogStreamState::StreamingPastLogs {
+                    poll_interval,
+                    filter,
+                    past_logs,
+                } => {
+                    if let Some(log) = past_logs.pop() {
+                        return Poll::Ready(Some(Ok(log)));
+                    } else {
+                        // TODO: this could duplicate some logs if they appear in both getFilterLogs
+                        // and getFilterChanges or it could skip some events.
```

Comment:
I would hope that `eth_getFilterLogs` would count as the "last polled" call when getting the `eth_getFilterChanges` :crossed_fingers: